examples/pytorch/compgcn/model.py

```diff
@@ -0,0 +1,371 @@
+        is a ('u' 'e' 'v') edge in the original graph, there must be a
+        ('v' 'e_inv' 'u') edge, where e_inv is mannual added to the original
+        graph AND edges in this type must have _inv in the end of edge name.
+
+        Parameters
+        ----------
+        g: DGL Graph
+            A DGL heterograph.
+
+        n_in_feats: dict[str, Tensor]
+            A dictionary of node type and node features.
+
+        e_in_feats: dict[str, Tensor]
+            A dictionary of edge type and edge features.
+
+        """
+        with g.local_scope():
+
+            # handle each edge type relation sub_graph
+            n_outputs = {ntype: [] for ntype in g.ntypes}
+
+            for stype, etype, dtype in g.canonical_etypes:
+
+                if etype not in e_in_feats:
+                    continue
+
+                # extract a subgraph in one edge type
+                rel_graph = g[stype, etype, dtype]
+
+                # check edge numbers
+                if rel_graph.number_of_edges() == 0:
+                    continue
+
+                # assign node and edge features to the subgraph in the name of 'h'
+                if stype not in n_in_feats or dtype not in n_in_feats:
+                    continue
+                else:
+                    rel_graph.nodes[stype].data['h'] = n_in_feats[stype]
+                    rel_graph.nodes[dtype].data['h'] = n_in_feats[dtype]
+
+                    rel_graph.edges[etype].data['h'] = e_in_feats[etype]
+
+                # compute norm values
+                in_norm = rel_graph.in_degrees().float().clamp(min=1)
+                in_norm = th.pow(in_norm, -0.5)
+
+                # compute composition function
+                if self.comp_fn == 'sub':
+                    rel_graph.update_all(dglfn.u_sub_e(lhs_field='h', rhs_field='h', out='m'),
+                                         dglfn.sum(msg='m', out='comp_h'))
+                elif self.comp_fn == 'mul':
+                    rel_graph.update_all(dglfn.u_mul_e(lhs_field='h', rhs_field='h', out='m'),
+                                         dglfn.sum(msg='m', out='comp_h'))
+                elif self.comp_fn == 'corr':
+                    rel_graph.update_all(lambda edges: {'corr_h': ccorr(edges.src['h'], edges.data['h'])},
+                                     dglfn.sum(msg='corr_h', out='comp_h'))
+                else:
+                    raise DGLError('Only supports sub, mul, and ccorr')
+
+                # normalize composition results
+                rel_graph.dstdata['comp_h'] = rel_graph.dstdata['comp_h'] * in_norm.view(-1, 1)
+
+                # linear transformation by different weights and add bias values
+                if str.endswith(etype, '_inv'):
+                    rel_graph.dstdata['comp_h'] = th.matmul(rel_graph.dstdata['comp_h'], self.W_O)
+                    if self.bias:
+                        rel_graph.dstdata['comp_h'] = rel_graph.dstdata['comp_h'] + self.b_O
```

Comment:
why not using the Pytorch linear projection layer?

Reply:
In the RGCN examples, they use Parameter instead of Linear. Not sure why, but I will change them to linear.